### PR TITLE
Constraints for LinearOperators

### PR DESCRIPTION
**Pull request by tamiko**
This is the current state of implementation where we stopped the discussion two
months ago.

### CODE REVIEW

File: include/deal.II/lac/constrained_linear_operator.h [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+ */
+template <typename Range, typename Domain>
+LinearOperator<Range, Domain> distribute_constraints_linear_operator(
+  const ConstraintMatrix &constraint_matrix,
+  const LinearOperator<Range, Domain> &exemplar)
+{
+  LinearOperator<Range, Domain> return_op = exemplar;
+
+  return_op.vmult_add = [&constraint_matrix](Range &v, const Domain &u)
+  {
```
Comment (tamiko) [inline include/deal.II/lac/constrained_linear_operator.h]: TODO: use intermediate storage if ```u``` and ```v``` are the same objects.

File: include/deal.II/lac/constrained_linear_operator.h [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+                v(i) +=  u(pos) * (*entries)[j].second;
+              }
+          }
+        else
+          v(i) += u(i);
+      }
+  };
+
+  return_op.Tvmult_add = [&constraint_matrix](Domain &v, const Range &u)
+  {
```
Comment (tamiko) [inline include/deal.II/lac/constrained_linear_operator.h]: TODO: use intermediate storage if ```u``` and ```v``` are the same objects.

**Comment by tamiko** [comment]
@bangerth I suggest that we keep this low level implementation for this pull-request.

I will have a look into improving the interface of ```ConstraintMatrix``` to directly support the distribution of linear constraints (for stored affine constraints). ```vmult``` should be straight-forward. The question is how to implement the transpose ```distribute``` that we need in an efficient manner.
All of this can be resolved in a follow-up pull-request.

File: doc/doxygen/headers/constraints.h [hunk 3]
```diff
@@ -432,4 +432,59 @@
...
+ * particular if there is no underlying matrix object that could be
+ * condensed (or taken care of constraints during assembly). This is
+ * usually the case if the system is described by a LinearOperator.
+ *
+ * In this case we can solve the modified system
+ * @f[
+ *   (C^T A C + Id_c) \tilde x = C^T (b - A\,k)
+ * @f]
+ * instead [1] (M. S. Shephard: Linear multipoint constraints applied via
+ * transformation as part of a direct stiffness assembly process, 1985).
```
Comment (bangerth) [inline doc/doxygen/headers/constraints.h]: can you complete the reference? It is missing journal, volume, pages.

File: doc/doxygen/headers/constraints.h [hunk 4]
```diff
@@ -432,4 +432,59 @@
...
+ *
+ * Here, $A$ is a given (unconstrained) system matrix and $b$
```
Comment (bangerth) [inline doc/doxygen/headers/constraints.h]: maybe say "...system matrix for which we only assume that we can apply it to a vector (but can not necessarily access individual matrix entries)"

File: include/deal.II/lac/constrained_linear_operator.h [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+ */
+//@{
+
+
+/**
+ * This function takes a ConstraintMatrix @p constraint_matrix and an
+ * operator exemplar @p exemplar (this exemplar is usually a linear
+ * operator that describes the system matrix) and returns a LinearOperator
+ * object associated with the "homogeneous action" of the underlying
+ * ConstraintMatrix object:
```
Comment (bangerth) [inline include/deal.II/lac/constrained_linear_operator.h]: I don't particularly care for the name "exemplar". An exemplar is not something to be used, but to be "imitated" or a "role model", see https://en.wiktionary.org/wiki/exemplar . But this function uses it to do very concrete things with. Maybe just call it `linear_operator` instead?

Comment (bangerth) [inline include/deal.II/lac/constrained_linear_operator.h]: Oh, scratch that -- I misread that this isn't actually the function that represents the constrained operator.

In that case, can you explain in the documentation why one needs to provide such an exemplar to begin with, and stress that the *action* of the exemplar is never used anywhere?

File: include/deal.II/lac/constrained_linear_operator.h [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+ *
+ * The LinearOperator object created by this function is primarily used
+ * internally in constrained_linear_operator() to build up a modified
+ * system of linear equations. How to solve a linear system of equations
+ * with this approach is explained in detail in the @ref constraints
+ * module.
+ *
+ * @author Mauro Bardelloni, Matthias Maier, 2015
+ *
+ * @relates LinearOperator
```
Comment (bangerth) [inline include/deal.II/lac/constrained_linear_operator.h]: Is it possible to add multiple `@relates` statements? In that case, maybe also relate it to `ConstraintMatrix`? Same for the other functions here (and possibly the whole namespace).

File: include/deal.II/lac/constrained_linear_operator.h (hunk 1 above)
Comment (bangerth) [inline include/deal.II/lac/constrained_linear_operator.h]: Feel free to add an assert if you don't want to implement it right away.

File: include/deal.II/lac/constrained_linear_operator.h [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+  };
+
+  return return_op;
+}
+
+
+/**
+ * Given a ConstraintMatrix @p constraint_matrix and an operator exemplar
+ * @p exemplar, return an LinearOperator that is the projection to the
+ * subspace of constrained degrees of freedom.
```
Comment (bangerth) [inline include/deal.II/lac/constrained_linear_operator.h]: an -> a

Can you describe in a formula what the returned object represents?

File: include/deal.II/lac/constrained_linear_operator.h [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+LinearOperator<Range, Domain>
+constrained_linear_operator(const ConstraintMatrix &constraint_matrix,
+                            const LinearOperator<Range, Domain> &linop)
+{
+  const auto C =
+    distribute_constraints_linear_operator(constraint_matrix, linop);
+  const auto Ct = transpose_operator(C);
+  const auto Id_c =
+    project_to_constrained_linear_operator(constraint_matrix, linop);
+  return Ct * linop * C + Id_c;
```
Comment (bangerth) [inline include/deal.II/lac/constrained_linear_operator.h]: nice code :-)

**Comment by bangerth** [comment]
Looks mostly good, up to just a few niggles.

File: include/deal.II/lac/constrained_linear_operator.h [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+            const auto *entries = constraint_matrix.get_constraint_entries (i);
+            for (types::global_dof_index j = 0; j < entries->size(); ++j)
+              {
+                const auto pos = (*entries)[j].first;
+                v(i) +=  u(pos) * (*entries)[j].second;
+              }
+          }
+        else
+          v(i) += u(i);
+      }
```
Comment (kronbichler) [inline include/deal.II/lac/constrained_linear_operator.h]: What is the assumption on the underlying vector type regarding parallel access? You loop through the locally owned indices of a vector `v` but later on you access more elements via `u(pos)`. Currently, the only vector type that supports this access pattern in parallel is parallel::distributed::Vector (if the ghost indices are based on the correct index set and if `update_ghost_values` has been called). One should probably add a warning about the fact that this is currently not implemented in parallel to help users identify the error message about access to non-owned entries at a somewhat unexpected place.

File: include/deal.II/lac/constrained_linear_operator.h (hunk 6 above)
Comment (tamiko) [inline include/deal.II/lac/constrained_linear_operator.h]: @bangerth Unfortunately, multiple @relates statements are not possible. (And this took me more than an hour to find out...)

File: include/deal.II/lac/constrained_linear_operator.h (hunk 9 above)
Comment (tamiko) [inline include/deal.II/lac/constrained_linear_operator.h]: @kronbichler I do not recall exactly any more. I think that above code was tested with the vector in ```TrilinosWrappers```, but to what generality I'm not sure.

I suggest that we fix this by implementing the necessary functionality (at least for ```vmult```) in ```ConstraintMatrix``` and call that.

Comment (tamiko) [inline include/deal.II/lac/constrained_linear_operator.h]: Maybe in a follow-up pull-request and clearly label that this object is not suitable for general use at the moment.

File: include/deal.II/lac/constrained_linear_operator.h (hunk 6 above)
Comment (bangerth) [inline include/deal.II/lac/constrained_linear_operator.h]: Thanks for checking. Feel free to adda link to this group of functions in the documentation of ConstraintMatrix if you find a convenient place (but not a requirement).

Comment (tamiko) [inline include/deal.II/lac/constrained_linear_operator.h]: > Thanks for checking. Feel free to adda link to this group of functions
> in the documentation of ConstraintMatrix if you find a convenient
> place (but not a requirement).

Will do that. Hopefully soon :-)

File: include/deal.II/lac/constrained_linear_operator.h (hunk 1 above)
Comment (tamiko) [inline include/deal.II/lac/constrained_linear_operator.h]: I have added an ```Assert```.

File: doc/news/changes.h [hunk 10]
```diff
@@ -183,6 +183,15 @@ inconvenience this causes.
 <a name="general"></a>
 <h3>General</h3>
 <ol>
+<<<<<<< 7f26aa71e926ca42536973cfb79514dc6d2321ff
```
Comment (kronbichler) [inline doc/news/changes.h]: Please remove this stray line

File: include/deal.II/lac/constrained_linear_operator.h (hunk 9 above)
Comment (kronbichler) [inline include/deal.II/lac/constrained_linear_operator.h]: This is a good plan.

**Comment by kronbichler** [comment]
Please fix that one line and merge yourself @tamiko

File: doc/news/changes.h (hunk 10 above)
Comment (tamiko) [inline doc/news/changes.h]: *ugh* Thanks for spotting this.

**Comment by luca-heltai** [comment]
:+1:

**Pull request merged**